diff --git a/src/Integrator/BDHI/BDHI_FCM.cu b/src/Integrator/BDHI/BDHI_FCM.cu
index 60d4044..fb56c04 100644
--- a/src/Integrator/BDHI/BDHI_FCM.cu
+++ b/src/Integrator/BDHI/BDHI_FCM.cu
@@ -4,7 +4,7 @@
 #include "BDHI_FCM.cuh"
 namespace uammd {
 namespace BDHI {
-auto FCMIntegrator::computeHydrodynamicDisplacements() {
+inline auto FCMIntegrator::computeHydrodynamicDisplacements() {
   auto pos = pd->getPos(access::location::gpu, access::mode::read);
   auto force = pd->getForce(access::location::gpu, access::mode::read);
   auto torque =
@@ -17,7 +17,7 @@ auto FCMIntegrator::computeHydrodynamicDisplacements() {
                                                temperature, 1.0 / sqrt(dt), st);
 }
 
-void FCMIntegrator::updateInteractors() {
+inline void FCMIntegrator::updateInteractors() {
   for (auto forceComp : interactors)
     forceComp->updateSimulationTime(steps * dt);
   if (steps == 1) {
@@ -29,7 +29,7 @@ void FCMIntegrator::updateInteractors() {
   }
 }
 
-void FCMIntegrator::resetForces() {
+inline void FCMIntegrator::resetForces() {
   int numberParticles = pg->getNumberParticles();
   auto force = pd->getForce(access::location::gpu, access::mode::write);
   auto forceGroup = pg->getPropertyIterator(force);
@@ -38,7 +38,7 @@ void FCMIntegrator::resetForces() {
   CudaCheckError();
 }
 
-void FCMIntegrator::resetTorques() {
+inline void FCMIntegrator::resetTorques() {
   int numberParticles = pg->getNumberParticles();
   auto torque = pd->getTorque(access::location::gpu, access::mode::write);
   auto torqueGroup = pg->getPropertyIterator(torque);
@@ -47,7 +47,7 @@ void FCMIntegrator::resetTorques() {
   CudaCheckError();
 }
 
-void FCMIntegrator::computeCurrentForces() {
+inline void FCMIntegrator::computeCurrentForces() {
   resetForces();
   if (pd->isDirAllocated())
     resetTorques();
@@ -92,7 +92,7 @@ __global__ void integrateEulerMaruyamaD(real4 *pos, real4 *dir,
 }
 } // namespace FCM_ns
 
-void FCMIntegrator::forwardTime() {
+inline void FCMIntegrator::forwardTime() {
   steps++;
   sys->log<System::DEBUG1>("[BDHI::FCM] Performing integration step %d", steps);
   updateInteractors();
diff --git a/src/Integrator/BDHI/BDHI_PSE.cuh b/src/Integrator/BDHI/BDHI_PSE.cuh
index c165fee..2547bb9 100644
--- a/src/Integrator/BDHI/BDHI_PSE.cuh
+++ b/src/Integrator/BDHI/BDHI_PSE.cuh
@@ -103,14 +103,14 @@ public:
     computeMFNearField(MF, st);
   }
 
-  void computeMFNearField(real3 *MF, cudaStream_t st) {
+  inline void computeMFNearField(real3 *MF, cudaStream_t st) {
     System::log<System::DEBUG1>("[BDHI::PSE] Computing MFNearField....");
     auto pd = pg->getParticleData();
     auto force = pd->getForce(access::location::gpu, access::mode::read);
     nearField->Mdot(force.begin(), MF, st);
   }
 
-  void computeMFFarField(real3 *MF, cudaStream_t st) {
+  inline void computeMFFarField(real3 *MF, cudaStream_t st) {
     System::log<System::DEBUG1>("[BDHI::PSE] Computing MFFarField....");
     auto pd = pg->getParticleData();
     int numberParticles = pg->getNumberParticles();
@@ -121,7 +121,7 @@ public:
                                                1.0 / sqrt(dt), st);
   }
 
-  void computeBdW(real3 *BdW, cudaStream_t st) {
+  inline void computeBdW(real3 *BdW, cudaStream_t st) {
     System::log<System::DEBUG2>("[BDHI::PSE] Real space brownian noise");
     // Far contribution is included in farField::Mdot
     nearField->computeStochasticDisplacements(BdW, temperature, 1.0, st);
@@ -139,7 +139,7 @@ public:
    *
    * The noise is scaled by the prefactor before adding to the result.
    */
-  void computeHydrodynamicDisplacements(real4 *force, real3 *MF,
+  inline void computeHydrodynamicDisplacements(real4 *force, real3 *MF,
                                         real temperature, real noise_prefactor,
                                         cudaStream_t st = 0) {
     int numberParticles = pg->getNumberParticles();
@@ -156,16 +156,16 @@ public:
                                                noise_prefactor, st);
   }
 
-  void setShearStrain(real newStrain) {
+  inline void setShearStrain(real newStrain) {
     System::log<System::DEBUG>("[BDHI::PSE] Shear strain changed to %g",
                                newStrain);
     nearField->setShearStrain(newStrain);
     farField->setShearStrain(newStrain);
   }
 
-  real getHydrodynamicRadius() { return hydrodynamicRadius; }
+  inline real getHydrodynamicRadius() { return hydrodynamicRadius; }
 
-  real getSelfMobility() { return this->M0; }
+  inline real getSelfMobility() { return this->M0; }
 
 private:
   shared_ptr<ParticleGroup> pg;
diff --git a/src/Integrator/BDHI/FCM/FCM_impl.cuh b/src/Integrator/BDHI/FCM/FCM_impl.cuh
index c632e2f..f5b8d87 100644
--- a/src/Integrator/BDHI/FCM/FCM_impl.cuh
+++ b/src/Integrator/BDHI/FCM/FCM_impl.cuh
@@ -97,9 +97,9 @@ public:
     cufftDestroy(cufft_plan_forward);
   }
 
-  real getHydrodynamicRadius() { return hydrodynamicRadius; }
-
-  real getSelfMobility() {
+  inline real getHydrodynamicRadius() { return hydrodynamicRadius; }
+  
+  inline real getSelfMobility() {
     // O(a^8) accuracy. See Hashimoto 1959.
     // With a Gaussian this expression has a minimum deviation from measuraments
     // of 7e-7*rh at L=64*rh. The translational invariance of the hydrodynamic
@@ -118,12 +118,12 @@ public:
            (1.0l - c * a + (4.0l / 3.0l) * M_PIl * a3 - a6pref * a3 * a3);
   }
 
-  Box getBox() { return this->box; }
+  inline Box getBox() { return this->box; }
 
   // Computes the velocities and angular velocities given the forces and torques
   //  If torques is a nullptr, the torque computation is skipped and the second
   //  output is empty
-  std::pair<cached_vector<real3>, cached_vector<real3>>
+  inline std::pair<cached_vector<real3>, cached_vector<real3>>
   computeHydrodynamicDisplacements(real4 *pos, real4 *force, real4 *torque,
                                    int numberParticles, real temperature,
                                    real prefactor, cudaStream_t st);
@@ -145,8 +145,8 @@ private:
   thrust::device_vector<char> cufftWorkArea;
 
   Parameters par;
-
-  void printMessages(Parameters par) {
+  
+  inline void printMessages(Parameters par) {
     auto rh = this->getHydrodynamicRadius();
     auto M0 = this->getSelfMobility();
     System::log<System::MESSAGE>("[BDHI::FCM] Using kernel: %s",
@@ -176,7 +176,7 @@ private:
     // %g", par.tolerance);
   }
 
-  void initCuFFT() {
+  inline void initCuFFT() {
     CufftSafeCall(cufftCreate(&cufft_plan_forward));
     CufftSafeCall(cufftCreate(&cufft_plan_inverse));
     CufftSafeCall(cufftSetAutoAllocation(cufft_plan_forward, 0));
@@ -290,7 +290,7 @@ auto interleave(Container &a, Container &b, Container &c) {
   return res;
 }
 
-cached_vector<complex3> forwardTransform(cached_vector<real3> &gridReal, int3 n,
+inline cached_vector<complex3> forwardTransform(cached_vector<real3> &gridReal, int3 n,
                                          cufftHandle plan, cudaStream_t st) {
   cached_vector<complex3> gridFourier((n.x / 2 + 1) * n.y * n.z);
   thrust::fill(thrust::cuda::par.on(st), gridFourier.begin(), gridFourier.end(),
@@ -303,7 +303,7 @@ cached_vector<complex3> forwardTransform(cached_vector<real3> &gridReal, int3 n,
   return gridFourier;
 }
 
-__global__ void addTorqueCurl(complex3 *gridTorquesFourier,
+__forceinline__ __global__ void addTorqueCurl(complex3 *gridTorquesFourier,
                               complex3 *gridVelsFourier, Grid grid) {
   int id = blockDim.x * blockIdx.x + threadIdx.x;
   const int3 nk = grid.cellDim;
@@ -372,7 +372,7 @@ void addSpreadTorquesFourier(IterPos &pos, IterTorque &torque,
  * \mathbf{k}^\mathbf{k} / \vert \mathbf{k} \vert^2$
  * @note This is executed with one thread per Fourier node.
  */
-__global__ void
+__forceinline__ __global__ void
 forceFourier2Vel(const complex3 *gridForces, /*Input array*/
                  complex3 *gridVels, /*Output array, can be the same as input*/
                  real vis, Grid grid /*Grid information and methods*/
@@ -396,7 +396,7 @@ forceFourier2Vel(const complex3 *gridForces, /*Input array*/
                  (B / real(ncells.x * ncells.y * ncells.z));
 }
 
-void convolveFourier(cached_vector<complex3> &gridVelsFourier, real viscosity,
+inline void convolveFourier(cached_vector<complex3> &gridVelsFourier, real viscosity,
                      Grid grid, cudaStream_t st) {
   System::log<System::DEBUG2>("[BDHI::FCM] Wave space velocity scaling");
   /*Scale the wave space grid forces, transforming in velocities -> B*FFT*S*F */
@@ -434,7 +434,7 @@ void convolveFourier(cached_vector<complex3> &gridVelsFourier, real viscosity,
  * Special care is taken to ensure that \f$v_k = v^*_{N-k}\f$, which implies
  * \f$dW_{w,k} = dW^*_{w, N-k}\f$.
  */
-__global__ void fourierBrownianNoise(complex3 *gridVelsFourier, Grid grid,
+__forceinline__ __global__ void fourierBrownianNoise(complex3 *gridVelsFourier, Grid grid,
                                      real prefactor, /* sqrt(2T/dt)*/
                                      real viscosity, uint seed1, uint seed2) {
   const uint id = blockIdx.x * blockDim.x + threadIdx.x;
@@ -511,7 +511,7 @@ __global__ void fourierBrownianNoise(complex3 *gridVelsFourier, Grid grid,
   }
 }
 
-void addBrownianNoise(cached_vector<complex3> &gridVelsFourier,
+inline void addBrownianNoise(cached_vector<complex3> &gridVelsFourier,
                       real temperature, real viscosity, real prefactor,
                       uint seed, Grid grid, cudaStream_t st) {
   static uint seed2 = 0;
@@ -541,7 +541,7 @@ void addBrownianNoise(cached_vector<complex3> &gridVelsFourier,
   }
 }
 
-cached_vector<real3> inverseTransform(cached_vector<complex3> &gridFourier,
+inline cached_vector<real3> inverseTransform(cached_vector<complex3> &gridFourier,
                                       int3 n, cufftHandle plan,
                                       cudaStream_t st) {
   int nx = 2 * (n.x / 2 + 1);
@@ -587,7 +587,7 @@ interpolateVelocity(IterPos &pos, cached_vector<real3> &gridVels, Grid grid,
 //  i*k_y*V_x)
 // Overwrite the output vector with the angular velocities in Fourier space
 // The input velocity vector is overwritten
-__global__ void computeVelocityCurlFourier(const complex3 *gridVelsFourier,
+__forceinline__ __global__ void computeVelocityCurlFourier(const complex3 *gridVelsFourier,
                                            complex3 *gridAngVelsFourier,
                                            Grid grid) {
   int id = blockDim.x * blockIdx.x + threadIdx.x;
@@ -613,7 +613,7 @@ __global__ void computeVelocityCurlFourier(const complex3 *gridVelsFourier,
   gridAngVelsFourier[id] = gridAng;
 }
 
-cached_vector<complex3>
+inline cached_vector<complex3>
 computeGridAngularVelocityFourier(cached_vector<complex3> &gridVelsFourier,
                                   Grid grid, cudaStream_t st) {
   const int3 n = grid.cellDim;
diff --git a/src/Integrator/BDHI/FCM/utils.cuh b/src/Integrator/BDHI/FCM/utils.cuh
index ca70d44..5ba5332 100644
--- a/src/Integrator/BDHI/FCM/utils.cuh
+++ b/src/Integrator/BDHI/FCM/utils.cuh
@@ -24,7 +24,7 @@ namespace fcm_detail {
 using complex = cufftComplex_t<real>;
 using complex3 = cufftComplex3_t<real>;
 
-__device__ int3 indexToWaveNumber(int i, int3 nk) {
+inline __device__ int3 indexToWaveNumber(int i, int3 nk) {
   int ikx = i % (nk.x / 2 + 1);
   int iky = (i / (nk.x / 2 + 1)) % nk.y;
   int ikz = i / ((nk.x / 2 + 1) * nk.y);
@@ -34,11 +34,11 @@ __device__ int3 indexToWaveNumber(int i, int3 nk) {
   return make_int3(ikx, iky, ikz);
 }
 
-__device__ real3 waveNumberToWaveVector(int3 ik, real3 L) {
+inline __device__ real3 waveNumberToWaveVector(int3 ik, real3 L) {
   return (real(2.0) * real(M_PI) / L) * make_real3(ik.x, ik.y, ik.z);
 }
 
-__device__ real3 getGradientFourier(int3 ik, int3 nk, real3 L) {
+inline __device__ real3 getGradientFourier(int3 ik, int3 nk, real3 L) {
   const bool isUnpairedX = ik.x == (nk.x - ik.x);
   const bool isUnpairedY = ik.y == (nk.y - ik.y);
   const bool isUnpairedZ = ik.z == (nk.z - ik.z);
@@ -67,7 +67,7 @@ __device__ real3 getGradientFourier(int3 ik, int3 nk, real3 L) {
  * @param fr The vector to be projected.
  * @return The projected vector \f$\mathbf{res}\f$.
  */
-__device__ real3 projectFourier(real k2, real3 dk, real3 fr) {
+inline __device__ real3 projectFourier(real k2, real3 dk, real3 fr) {
   const real invk2 = real(1.0) / k2;
   real3 vr = fr - dk * dot(fr, dk * invk2);
   return vr;
@@ -90,7 +90,7 @@ __device__ real3 projectFourier(real k2, real3 dk, real3 fr) {
  * @param fr The complex-valued vector to be projected.
  * @return The projected complex vector \f$\mathbf{res}\f$.
  */
-__device__ complex3 projectFourier(real k2, real3 dk, complex3 factor) {
+inline __device__ complex3 projectFourier(real k2, real3 dk, complex3 factor) {
   real3 re =
       projectFourier(k2, dk, make_real3(factor.x.x, factor.y.x, factor.z.x));
   real3 imag =
@@ -112,7 +112,7 @@ __device__ complex3 projectFourier(real k2, real3 dk, complex3 factor) {
  * @param prefactor The target variance magnitude for the noise.
  * @return A complex-valued noise vector \f$dW\f$ with the specified statistics.
  */
-__device__ complex3 generateNoise(
+inline __device__ complex3 generateNoise(
     real prefactor, uint id, uint seed1,
     uint seed2) { // Uncomment to use uniform numbers instead of gaussian
   Saru saru(id, seed1, seed2);
@@ -129,7 +129,7 @@ __device__ complex3 generateNoise(
   return noise;
 }
 
-__device__ bool isNyquistWaveNumber(int3 cell, int3 ncells) {
+inline __device__ bool isNyquistWaveNumber(int3 cell, int3 ncells) {
   /*Beware of nyquist points! They only appear with even cell dimensions
     There are 8 nyquist points at most (cell=0,0,0 is excluded at the start of
     the kernel) These are the 8 vertex of the inferior left cuadrant. The O
diff --git a/src/Integrator/VerletNVE.cu b/src/Integrator/VerletNVE.cu
index b37d22f..0df8345 100644
--- a/src/Integrator/VerletNVE.cu
+++ b/src/Integrator/VerletNVE.cu
@@ -29,7 +29,7 @@
 
 namespace uammd {
 
-VerletNVE::VerletNVE(shared_ptr<ParticleGroup> pg, VerletNVE::Parameters par)
+inline VerletNVE::VerletNVE(shared_ptr<ParticleGroup> pg, VerletNVE::Parameters par)
     : Integrator(pg, "VerletNVE"), dt(par.dt), energy(par.energy),
       is2D(par.is2D), initVelocities(par.initVelocities), steps(0) {
   if (initVelocities)
@@ -56,7 +56,7 @@ VerletNVE::VerletNVE(shared_ptr<ParticleGroup> pg, VerletNVE::Parameters par)
   CudaSafeCall(cudaStreamCreate(&stream));
 }
 
-VerletNVE::~VerletNVE() { cudaStreamDestroy(stream); }
+inline VerletNVE::~VerletNVE() { cudaStreamDestroy(stream); }
 
 namespace VerletNVE_ns {
 
@@ -85,7 +85,7 @@ __global__ void integrateGPU(real4 *__restrict__ pos, real3 *__restrict__ vel,
 }
 } // namespace VerletNVE_ns
 
-void VerletNVE::initializeVelocities() {
+inline void VerletNVE::initializeVelocities() {
   // In the first step, compute energy in the system
   // in order to adapt the initial kinetic energy to match the input total
   // energy E = U+K
@@ -149,7 +149,7 @@ template <int step> void VerletNVE::callIntegrate() {
       numberParticles, dt, is2D);
 }
 
-void VerletNVE::resetForces() {
+inline void VerletNVE::resetForces() {
   int numberParticles = pg->getNumberParticles();
   auto force = pd->getForce(access::location::gpu, access::mode::write);
   auto force_gr = pg->getPropertyIterator(force);
@@ -157,7 +157,7 @@ void VerletNVE::resetForces() {
                force_gr + numberParticles, real4());
 }
 
-void VerletNVE::firstStepPreparation() {
+inline void VerletNVE::firstStepPreparation() {
   if (initVelocities) {
     initializeVelocities();
   }
@@ -171,7 +171,7 @@ void VerletNVE::firstStepPreparation() {
 }
 
 // Move the particles in my group 1 dt in time.
-void VerletNVE::forwardTime() {
+inline void VerletNVE::forwardTime() {
   steps++;
   sys->log<System::DEBUG1>("[VerletNVE] Performing integration step %d", steps);
   if (steps == 1)
@@ -200,7 +200,7 @@ __global__ void sumKineticEnergy(const VelocityIterator vel,
 }
 }; // namespace verletnve_ns
 
-real VerletNVE::sumEnergy() {
+inline real VerletNVE::sumEnergy() {
   int numberParticles = pg->getNumberParticles();
   auto vel_gr = pg->getPropertyIterator(
       pd->getVel(access::location::gpu, access::mode::read));
diff --git a/src/Interactor/NeighbourList/BasicList/BasicListBase.cuh b/src/Interactor/NeighbourList/BasicList/BasicListBase.cuh
index da1fcf9..3da64ec 100644
--- a/src/Interactor/NeighbourList/BasicList/BasicListBase.cuh
+++ b/src/Interactor/NeighbourList/BasicList/BasicListBase.cuh
@@ -39,7 +39,7 @@ list is to be used several times between updates.
 
 namespace uammd {
 namespace BasicNeighbourList_ns {
-__global__ void fillBasicNeighbourList(CellList_ns::NeighbourContainer ni,
+__forceinline__ __global__ void fillBasicNeighbourList(CellList_ns::NeighbourContainer ni,
                                        int *neighbourList,
                                        int *numberNeighbours,
                                        int maxNeighboursPerParticle,
@@ -84,7 +84,7 @@ protected:
   int maxNeighboursPerParticle;
 
   struct NeighbourListOffsetFunctor {
-    NeighbourListOffsetFunctor(int str) : stride(str) {}
+    inline NeighbourListOffsetFunctor(int str) : stride(str) {}
     int stride;
     inline __host__ __device__ int operator()(const int &index) const {
       // return index*stride;
@@ -96,7 +96,7 @@ protected:
   using StrideIterator = thrust::transform_iterator<NeighbourListOffsetFunctor,
                                                     CountingIterator, int>;
 
-  Grid createUpdateGrid(Box box, real cutOff) {
+  inline Grid createUpdateGrid(Box box, real cutOff) {
     real3 L = box.boxSize;
     constexpr real inf = std::numeric_limits<real>::max();
     // If the box is non periodic L and cellDim are free parameters
@@ -125,12 +125,12 @@ protected:
   }
 
 public:
-  BasicNeighbourListBase() {
+  inline BasicNeighbourListBase() {
     maxNeighboursPerParticle = 32;
     errorFlags.resize(1);
   }
 
-  ~BasicNeighbourListBase() {}
+  inline ~BasicNeighbourListBase() {}
 
   template <class PositionIterator>
   void update(PositionIterator pos, int numberParticles, Box box, real cutOff,
@@ -153,7 +153,7 @@ public:
         StrideIterator(CountingIterator(0), NeighbourListOffsetFunctor(0));
   };
 
-  BasicNeighbourListData getBasicNeighbourList(cudaStream_t st = 0) {
+  inline BasicNeighbourListData getBasicNeighbourList(cudaStream_t st = 0) {
     BasicNeighbourListData nl;
     auto cld = cl.getCellList();
     nl.neighbourList = thrust::raw_pointer_cast(neighbourList.data());
@@ -169,18 +169,18 @@ public:
   }
 
 private:
-  void resizeNeighbourListToCurrent(int numberParticles) {
+  inline void resizeNeighbourListToCurrent(int numberParticles) {
     neighbourList.resize(numberParticles * (maxNeighboursPerParticle + 1));
     numberNeighbours.resize(numberParticles);
   }
 
-  void fillBasicNeighbourList(cudaStream_t st) {
+  inline void fillBasicNeighbourList(cudaStream_t st) {
     while (not tryToFillNeighbourList(st)) {
       increaseMaximumNeighboursPerParticle();
     }
   }
 
-  bool tryToFillNeighbourList(cudaStream_t st) {
+  inline bool tryToFillNeighbourList(cudaStream_t st) {
     System::log<System::DEBUG3>(
         "[BasicList] Attempting to fill list with %d neighbours per particle",
         maxNeighboursPerParticle);
@@ -205,7 +205,7 @@ private:
     return not foundTooManyNeighbours;
   }
 
-  void increaseMaximumNeighboursPerParticle() {
+  inline void increaseMaximumNeighboursPerParticle() {
     this->maxNeighboursPerParticle += 32;
     System::log<System::DEBUG3>(
         "[BasicList] Increasing maximum number of neighbours to %d",
diff --git a/src/Interactor/Potential/DPD.cuh b/src/Interactor/Potential/DPD.cuh
index f4f238f..f20688c 100644
--- a/src/Interactor/Potential/DPD.cuh
+++ b/src/Interactor/Potential/DPD.cuh
@@ -158,7 +158,7 @@ public:
     }
   };
 
-  ForceTransverser getForceTransverser(Box box, shared_ptr<ParticleData> pd) {
+  inline ForceTransverser getForceTransverser(Box box, shared_ptr<ParticleData> pd) {
     auto pos = pd->getPos(access::location::gpu, access::mode::read);
     auto vel = pd->getVel(access::location::gpu, access::mode::read);
     auto force = pd->getForce(access::location::gpu, access::mode::readwrite);
@@ -173,14 +173,14 @@ public:
   // the method getEnergyTransverser is not found and the energy will not be
   // computed altogether.
 
-  BasicNullTransverser getForceEnergyTransverser(Box box,
+  inline BasicNullTransverser getForceEnergyTransverser(Box box,
                                                  shared_ptr<ParticleData> pd) {
     System::log<System::CRITICAL>("[DPD] No way of measuring energy in DPD");
     return BasicNullTransverser();
   }
 };
 
-template <> void DPD_impl<DefaultDissipation>::printGamma() {
+template <> inline void DPD_impl<DefaultDissipation>::printGamma() {
   System::log<System::MESSAGE>("[Potential::DPD] gamma: %f", gamma.gamma);
 }
 
diff --git a/src/ParticleData/ParticleData.cuh b/src/ParticleData/ParticleData.cuh
index b7401d8..89bb8b5 100644
--- a/src/ParticleData/ParticleData.cuh
+++ b/src/ParticleData/ParticleData.cuh
@@ -468,7 +468,7 @@ private:
 #define INIT_PROPERTIES(r, data, tuple)                                        \
   INIT_PROPERTIES_T(PROPNAME_CAPS(tuple), PROPNAME(tuple))
 
-ParticleData::ParticleData(int numberParticles, shared_ptr<System> sys)
+inline ParticleData::ParticleData(int numberParticles, shared_ptr<System> sys)
     : numberParticles(numberParticles), originalOrderIndexCPUNeedsUpdate(true),
       sys(sys) PROPERTY_LOOP(INIT_PROPERTIES) {
   if (!sys) {
@@ -489,7 +489,7 @@ ParticleData::ParticleData(int numberParticles, shared_ptr<System> sys)
   particle_sorter = std::make_shared<ParticleSorter>();
 }
 
-void ParticleData::sortParticles(cudaStream_t st) {
+inline void ParticleData::sortParticles(cudaStream_t st) {
   sys->log<System::DEBUG>("[ParticleData] Sorting particles...");
 
   {
@@ -521,7 +521,7 @@ void ParticleData::sortParticles(cudaStream_t st) {
   this->emitReorder();
 }
 
-void ParticleData::changeNumParticles(int Nnew) {
+inline void ParticleData::changeNumParticles(int Nnew) {
   sys->log<System::CRITICAL>(
       "[ParticleData] CHANGE PARTICLES FUNCTIONALITY NOT IMPLEMENTED YET!!!");
   sys->log<System::DEBUG>("[ParticleData] Adding/Removing particles...");
diff --git a/src/ParticleData/ParticleGroup.cuh b/src/ParticleData/ParticleGroup.cuh
index 73b7260..a855e0e 100644
--- a/src/ParticleData/ParticleGroup.cuh
+++ b/src/ParticleData/ParticleGroup.cuh
@@ -137,7 +137,7 @@ public:
 namespace ParticleGroup_ns {
 // Updates the indices of the particles in a group using
 // pd->getIdOrderedIndices()
-__global__ void updateGroupIndices( // An array that stores the indices of the
+__forceinline__ __global__ void updateGroupIndices( // An array that stores the indices of the
                                     // particles in the group per id.
     const int *__restrict__ id2index,
     // Out: the current ParticleData indices of the particles in the group
@@ -192,17 +192,17 @@ class ParticleGroup {
 
 public:
   // Defaults to all particles in group
-  ParticleGroup(std::shared_ptr<ParticleData> pd,
+  inline ParticleGroup(std::shared_ptr<ParticleData> pd,
                 std::string name = std::string("noName"));
 
   // Create the group from a selector
   template <class ParticleSelector>
-  ParticleGroup(ParticleSelector selector, std::shared_ptr<ParticleData> pd,
+  inline ParticleGroup(ParticleSelector selector, std::shared_ptr<ParticleData> pd,
                 std::string name = std::string("noName"));
 
   // Create the group from a list of particle IDs
   template <class InputIterator>
-  ParticleGroup(InputIterator begin, InputIterator end,
+  inline ParticleGroup(InputIterator begin, InputIterator end,
                 std::shared_ptr<ParticleData> pd,
                 std::string name = std::string("noName"));
 
@@ -226,7 +226,7 @@ public:
   void addParticlesByCurrentIndex(access::location loc, const int *indices,
                                   int N);
 
-  void handleReorder() {
+  inline void handleReorder() {
     sys->log<System::DEBUG>(
         "[ParticleGroup] Handling reorder signal in group %s",
         this->name.c_str());
@@ -236,7 +236,7 @@ public:
     }
   }
 
-  const int *getGroupIndexMaskRawPtr(access::location loc) {
+  inline const int *getGroupIndexMaskRawPtr(access::location loc) {
     int *ptr;
 
     if (this->allParticlesInGroup and updateMask) {
@@ -277,7 +277,7 @@ public:
   }
 
   // Get a raw memory pointer to the index list if it exists
-  const int *getIndicesRawPtr(access::location loc) {
+  const inline int *getIndicesRawPtr(access::location loc) {
     if (this->allParticlesInGroup || numberParticles == 0)
       return nullptr;
     this->computeIndexList();
@@ -319,7 +319,7 @@ public:
   using accessIterator = thrust::permutation_iterator<Iterator, IndexIterator>;
 
   // Get an iterator with the indices of particles in this group
-  IndexIterator getIndexIterator(access::location loc) {
+  inline IndexIterator getIndexIterator(access::location loc) {
     return IndexIterator(thrust::make_counting_iterator<int>(0),
                          IndexAccess(this->getIndicesRawPtr(loc)));
   }
@@ -342,7 +342,7 @@ public:
   // in that index is
   //  part of the group and 0 otherwise.
   // loc specifies the memory location of the underlying array
-  MaskIterator getGroupIndexMask(access::location loc) {
+  inline MaskIterator getGroupIndexMask(access::location loc) {
     return MaskIterator(thrust::make_counting_iterator<int>(0),
                         MaskAccess(this->getGroupIndexMaskRawPtr(loc)));
   }
@@ -385,7 +385,7 @@ public:
 };
 
 template <class ParticleSelector>
-ParticleGroup::ParticleGroup(ParticleSelector selector,
+inline ParticleGroup::ParticleGroup(ParticleSelector selector,
                              std::shared_ptr<ParticleData> pd, std::string name)
     : pd(pd), sys(pd->getSystem()), name(name) {
   sys->log<System::MESSAGE>("[ParticleGroup] Group %s created with selector %s",
@@ -413,7 +413,7 @@ ParticleGroup::ParticleGroup(ParticleSelector selector,
 
 // Specialization of a particle group with an All selector
 template <>
-ParticleGroup::ParticleGroup(particle_selector::All selector,
+inline ParticleGroup::ParticleGroup(particle_selector::All selector,
                              std::shared_ptr<ParticleData> pd, std::string name)
     : pd(pd), sys(pd->getSystem()), name(name) {
   sys->log<System::MESSAGE>(
@@ -426,12 +426,12 @@ ParticleGroup::ParticleGroup(particle_selector::All selector,
                             name.c_str(), numberParticles);
 }
 
-ParticleGroup::ParticleGroup(std::shared_ptr<ParticleData> pd, std::string name)
+inline ParticleGroup::ParticleGroup(std::shared_ptr<ParticleData> pd, std::string name)
     : ParticleGroup(particle_selector::All(), pd, name) {}
 
 // Specialization of an empty particle group
 template <>
-ParticleGroup::ParticleGroup(particle_selector::None selector,
+inline ParticleGroup::ParticleGroup(particle_selector::None selector,
                              std::shared_ptr<ParticleData> pd, std::string name)
     : pd(pd), sys(pd->getSystem()), name(name) {
   this->allParticlesInGroup = false;
@@ -470,7 +470,7 @@ ParticleGroup::ParticleGroup(InputIterator begin, InputIterator end,
 // This is trivial with  pd->getIdOrderedIndices()!
 // Handle a reordering of the particles (which invalids the previous relation
 // between IDs and indices)
-void ParticleGroup::computeIndexList(bool forceUpdate) {
+inline void ParticleGroup::computeIndexList(bool forceUpdate) {
   if (numberParticles == 0)
     return;
   if (this->needsIndexListUpdate || forceUpdate) { // Update only if needed
@@ -495,7 +495,7 @@ void ParticleGroup::computeIndexList(bool forceUpdate) {
   }
 }
 // Add particles to the group via an array with ids
-void ParticleGroup::addParticlesById(access::location loc, const int *ids,
+inline void ParticleGroup::addParticlesById(access::location loc, const int *ids,
                                      int N) {
   sys->log<System::DEBUG1>(
       "[ParticleGroup] Adding %d particles to group %s via ids!", N,
@@ -536,7 +536,7 @@ void ParticleGroup::addParticlesById(access::location loc, const int *ids,
 }
 
 namespace ParticleGroup_ns {
-__global__ void IdsFromIndices(const int *indices, const int *index2Id,
+__forceinline__ __global__ void IdsFromIndices(const int *indices, const int *index2Id,
                                int *groupParticleIds, int N) {
   int tid = blockIdx.x * blockDim.x + threadIdx.x;
   if (tid >= N)
@@ -549,7 +549,7 @@ __global__ void IdsFromIndices(const int *indices, const int *index2Id,
 } // namespace ParticleGroup_ns
 // Add particles to the group via an array with the current indices of the
 // particles in pd (faster)
-void ParticleGroup::addParticlesByCurrentIndex(access::location loc,
+inline void ParticleGroup::addParticlesByCurrentIndex(access::location loc,
                                                const int *indices, int N) {
   sys->log<System::DEBUG1>(
       "[ParticleGroup] Adding %d particles to group %s via indices!", N,
diff --git a/src/System/Log.h b/src/System/Log.h
index d16fd3f..448e95a 100644
--- a/src/System/Log.h
+++ b/src/System/Log.h
@@ -34,7 +34,7 @@ constexpr int maxLogLevel = MAXLOGLEVEL;
 constexpr int maxLogLevel = 6;
 #endif
 
-ElementType getLogLevelInfo(int level) {
+inline ElementType getLogLevelInfo(int level) {
   static const std::map<int, ElementType> printMap{
       {CRITICAL, std::make_tuple(stderr, "\e[101m[CRITICAL] ", "\e[0m\n")},
       {ERROR, std::make_tuple(stderr, "\e[91m[ERROR] \e[0m", "\n")},
diff --git a/src/System/System.h b/src/System/System.h
index ad84d32..c8105bc 100644
--- a/src/System/System.h
+++ b/src/System/System.h
@@ -315,7 +315,7 @@ public:
   }
 };
 
-void System::printWelcome() {
+inline void System::printWelcome() {
   std::string separator;
   fori(0, 60) separator += "━";
   separator += "┓";
@@ -339,7 +339,7 @@ void System::printWelcome() {
   log<System::MESSAGE>("Computation started at %s", std::ctime(&time));
 }
 
-void System::printFarewell() {
+inline void System::printFarewell() {
   cudaDeviceSynchronize();
   std::time_t time =
       std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
diff --git a/src/misc/BoundaryValueProblem/BVPMemory.cuh b/src/misc/BoundaryValueProblem/BVPMemory.cuh
index 7daf35c..583e591 100644
--- a/src/misc/BoundaryValueProblem/BVPMemory.cuh
+++ b/src/misc/BoundaryValueProblem/BVPMemory.cuh
@@ -18,7 +18,7 @@ template <class T> struct StorageHandle {
 };
 
 namespace detail {
-size_t computeExtraAlignment(size_t addressOffset, size_t alignment) {
+inline size_t computeExtraAlignment(size_t addressOffset, size_t alignment) {
   size_t extraAlignment = 0;
   bool isMissAligned = addressOffset % alignment != 0;
   if (isMissAligned)
@@ -122,4 +122,4 @@ public:
 
 } // namespace BVP
 } // namespace uammd
-#endif
\ No newline at end of file
+#endif
diff --git a/src/misc/IBM_kernels.cuh b/src/misc/IBM_kernels.cuh
index ba3b84c..019f00a 100644
--- a/src/misc/IBM_kernels.cuh
+++ b/src/misc/IBM_kernels.cuh
@@ -80,7 +80,7 @@ template <class Foo> auto integrate(Foo foo, real rmin, real rmax, int Nr) {
 } // namespace detail
 
 //[1] Taken from https://arxiv.org/pdf/1712.04732.pdf
-__host__ __device__ real BM(real zz, real alpha, real beta) {
+inline __host__ __device__ real BM(real zz, real alpha, real beta) {
   const real z = zz / alpha;
   const real z2 = z * z;
   const real dz2 = real(1.0) - z2;
diff --git a/src/misc/LanczosAlgorithm.cuh b/src/misc/LanczosAlgorithm.cuh
index ffd371a..cd024a8 100644
--- a/src/misc/LanczosAlgorithm.cuh
+++ b/src/misc/LanczosAlgorithm.cuh
@@ -43,7 +43,7 @@ struct Solver {
           cudaStream_t st = 0) {
     return run(&dot, Bv, v, tolerance, N, st);
   }
-  int run(std::function<void(real *, real *)> dot, real *Bv, const real *v,
+  inline int run(std::function<void(real *, real *)> dot, real *Bv, const real *v,
           real tolerance, int N, cudaStream_t st = 0) {
     auto lanczos_dot = createMatrixDotAdaptor(dot);
     return run(lanczos_dot, Bv, v, tolerance, N, st);
@@ -53,21 +53,21 @@ struct Solver {
   // iterations B = sqrt(M)
   real runIterations(MatrixDot *dot, real *Bz, const real *z,
                      int numberIterations, int N);
-  real runIterations(MatrixDot &dot, real *Bv, const real *v,
+  inline real runIterations(MatrixDot &dot, real *Bv, const real *v,
                      int numberIterations, int N) {
     return runIterations(&dot, Bv, v, numberIterations, N);
   }
-  real runIterations(std::function<void(real *, real *)> dot, real *Bv,
+  inline real runIterations(std::function<void(real *, real *)> dot, real *Bv,
                      const real *v, int numberIterations, int N) {
     auto lanczos_dot = createMatrixDotAdaptor(dot);
     return runIterations(lanczos_dot, Bv, v, numberIterations, N);
   }
 
-  void setIterationHardLimit(int newLimit) {
+  inline void setIterationHardLimit(int newLimit) {
     this->iterationHardLimit = newLimit;
   }
 
-  int getLastRunRequiredSteps() { return this->lastRunRequiredSteps; }
+  inline int getLastRunRequiredSteps() { return this->lastRunRequiredSteps; }
 
 private:
   // Increases storage space
diff --git a/src/misc/LanczosAlgorithm/LanczosAlgorithm.cu b/src/misc/LanczosAlgorithm/LanczosAlgorithm.cu
index 0486d2d..8406d7f 100644
--- a/src/misc/LanczosAlgorithm/LanczosAlgorithm.cu
+++ b/src/misc/LanczosAlgorithm/LanczosAlgorithm.cu
@@ -40,7 +40,7 @@ class KrylovSubspace {
   int subSpaceDimension;
 
   real normz;
-  void diagonalizeSubSpace() {
+inline void diagonalizeSubSpace() {
     int size = getSubSpaceSize();
     /**************LAPACKE********************/
     /*The tridiagonal matrix is stored only with its diagonal and subdiagonal*/
@@ -62,7 +62,7 @@ class KrylovSubspace {
     }
   }
 
-  real *computeSquareRoot() {
+  inline real *computeSquareRoot() {
     int size = getSubSpaceSize();
     diagonalizeSubSpace();
     /***Hdiag_temp = Hdiag*P*e1****/
@@ -81,9 +81,9 @@ class KrylovSubspace {
     return detail::getRawPointer(htempGPU);
   }
 
-  real *getTransformationMatrix() { return detail::getRawPointer(V); }
+  inline real *getTransformationMatrix() { return detail::getRawPointer(V); }
 
-  void resize(int subSpaceSize) {
+  inline void resize(int subSpaceSize) {
     CudaSafeCall(cudaDeviceSynchronize());
     try {
       w.resize((N + 1), real());
@@ -99,16 +99,16 @@ class KrylovSubspace {
   }
 
 public:
-  KrylovSubspace(int N) : subSpaceDimension(0), N(N) {
+  inline KrylovSubspace(int N) : subSpaceDimension(0), N(N) {
     // Init cuBLAS for Lanczos process
     CublasSafeCall(cublasCreate_v2(&cublas_handle));
     this->resize(1);
   }
 
-  ~KrylovSubspace() { CublasSafeCall(cublasDestroy(cublas_handle)); }
+  inline ~KrylovSubspace() { CublasSafeCall(cublasDestroy(cublas_handle)); }
 
   /************v[0] = z/||z||_2*****/
-  void setFirstBasisVector(const real *z) {
+  inline void setFirstBasisVector(const real *z) {
     /*1/norm(z)*/
     real *Vm = getTransformationMatrix();
     lanczos::device_nrm2(cublas_handle, N, z, 1, &(this->normz));
@@ -120,7 +120,7 @@ public:
     lanczos::device_scal(cublas_handle, N, &invz2, Vm, 1);
   }
 
-  void nextIteration(MatrixDot *dot) {
+  inline void nextIteration(MatrixDot *dot) {
     int i = subSpaceDimension;
     this->subSpaceDimension++;
     resize(subSpaceDimension + 1);
@@ -156,10 +156,10 @@ public:
     detail::device_copy(w.begin(), w.begin() + N, V.begin() + N * (i + 1));
   }
 
-  int getSubSpaceSize() { return subSpaceDimension; }
+  inline int getSubSpaceSize() { return subSpaceDimension; }
 
   // Computes the current result guess sqrt(M)*v, stores in BdW
-  void computeCurrentResultEstimation(real *BdW) {
+  inline void computeCurrentResultEstimation(real *BdW) {
     int m = getSubSpaceSize();
     /**** y = ||z||_2 * Vm * H^1/2 * e_1 *****/
     /**** H^1/2*e1 = Pt* first_column_of(sqrt(Hdiag)*P) ******/
@@ -173,14 +173,14 @@ public:
 };
 } // namespace detail
 
-Solver::Solver() : check_convergence_steps(3) {
+inline Solver::Solver() : check_convergence_steps(3) {
   // Init cuBLAS for Lanczos process
   CublasSafeCall(cublasCreate_v2(&cublas_handle));
 }
 
-Solver::~Solver() { CublasSafeCall(cublasDestroy(cublas_handle)); }
+inline Solver::~Solver() { CublasSafeCall(cublasDestroy(cublas_handle)); }
 
-real Solver::runIterations(MatrixDot *dot, real *Bz, const real *z,
+inline real Solver::runIterations(MatrixDot *dot, real *Bz, const real *z,
                            int numberIterations, int N) {
   oldBz.resize((N + 1), real());
   /*Lanczos iterations for Krylov decomposition*/
@@ -199,7 +199,7 @@ real Solver::runIterations(MatrixDot *dot, real *Bz, const real *z,
   return currentResidual;
 }
 
-int Solver::run(MatrixDot *dot, real *Bz, const real *z, real tolerance, int N,
+inline int Solver::run(MatrixDot *dot, real *Bz, const real *z, real tolerance, int N,
                 cudaStream_t st) {
   oldBz.resize((N + 1));
   detail::device_fill(oldBz.begin(), oldBz.end(), real());
@@ -227,7 +227,7 @@ int Solver::run(MatrixDot *dot, real *Bz, const real *z, real tolerance, int N,
   throw std::runtime_error("[Lanczos] Could not converge");
 }
 
-real Solver::computeError(real *Bz, int N, int iter) {
+inline real Solver::computeError(real *Bz, int N, int iter) {
   /*Compute error as in eq 27 in [1]
     Error = ||Bz_i - Bz_{i-1}||_2 / ||Bz_{i-1}||_2
   */
@@ -250,7 +250,7 @@ real Solver::computeError(real *Bz, int N, int iter) {
   return Error;
 }
 
-void Solver::registerRequiredStepsForConverge(int steps_needed) {
+inline void Solver::registerRequiredStepsForConverge(int steps_needed) {
   this->lastRunRequiredSteps = steps_needed;
   if (steps_needed - 2 > check_convergence_steps) {
     check_convergence_steps += 1;
diff --git a/src/utils/Grid.cuh b/src/utils/Grid.cuh
index 2b9c0fd..a7c7207 100644
--- a/src/utils/Grid.cuh
+++ b/src/utils/Grid.cuh
@@ -139,7 +139,7 @@ struct Grid {
 
 // Looks for the closest (equal or greater) number of nodes of the form
 // 2^a*3^b*5^c*7^d*11^e
-int3 nextFFTWiseSize3D(int3 size) {
+inline int3 nextFFTWiseSize3D(int3 size) {
   int *cdim = &size.x;
   using integral = uint64_t;
   integral max_dim = std::max({size.x, size.y, size.z});
diff --git a/src/utils/ParticleSorter.cuh b/src/utils/ParticleSorter.cuh
index 64cbdd4..e00152a 100644
--- a/src/utils/ParticleSorter.cuh
+++ b/src/utils/ParticleSorter.cuh
@@ -90,7 +90,7 @@ struct CellIndexHash {
   }
 };
 
-int clz(uint n) {
+inline int clz(uint n) {
   n |= (n >> 1);
   n |= (n >> 2);
   n |= (n >> 4);
@@ -100,7 +100,7 @@ int clz(uint n) {
 }
 
 template <class HashIterator>
-__global__ void assignHash(HashIterator hasher, int *__restrict__ index,
+__forceinline__ __global__ void assignHash(HashIterator hasher, int *__restrict__ index,
                            uint *__restrict__ hash, int N) {
   const int i = blockIdx.x * blockDim.x + threadIdx.x;
   if (i >= N)
@@ -219,7 +219,7 @@ private:
   thrust::device_vector<int> index, index_alt;
   thrust::device_vector<uint> hash, hash_alt;
 
-  void tryToUpdateOrderById(int *id, int N, cudaStream_t st) {
+  inline void tryToUpdateOrderById(int *id, int N, cudaStream_t st) {
     original_index.resize(N);
     index_alt.resize(N);
     hash.resize(N);
diff --git a/src/utils/cublasDebug.h b/src/utils/cublasDebug.h
index caf307c..0cec3f7 100644
--- a/src/utils/cublasDebug.h
+++ b/src/utils/cublasDebug.h
@@ -7,7 +7,7 @@
 #include <cublas_v2.h>
 #define CublasSafeCall(err) __cublasSafeCall(err, __FILE__, __LINE__)
 
-const char *cublasGetErrorString(cublasStatus_t status) {
+inline const char *cublasGetErrorString(cublasStatus_t status) {
   switch (status) {
   case CUBLAS_STATUS_SUCCESS:
     return "CUBLAS_STATUS_SUCCESS";
diff --git a/src/utils/cufftDebug.h b/src/utils/cufftDebug.h
index 945d320..b2967fc 100644
--- a/src/utils/cufftDebug.h
+++ b/src/utils/cufftDebug.h
@@ -12,7 +12,7 @@
 #define CufftSafeCall(err) __cufftSafeCall(err, __FILE__, __LINE__)
 
 namespace uammd {
-const char *cufftGetErrorString(cufftResult_t err) {
+inline const char *cufftGetErrorString(cufftResult_t err) {
   switch (err) {
   case CUFFT_INVALID_PLAN:
     return "CUFFT_INVALID_PLAN\n";
diff --git a/src/utils/cufftPrecisionAgnostic.h b/src/utils/cufftPrecisionAgnostic.h
index 9ebc0f3..63c7b90 100644
--- a/src/utils/cufftPrecisionAgnostic.h
+++ b/src/utils/cufftPrecisionAgnostic.h
@@ -46,14 +46,14 @@ cufftResult cufftExecReal2Complex(cufftHandle &plan, cufftReal_t<prec> *d_in,
                                   cufftComplex_t<prec> *d_out);
 
 template <>
-cufftResult cufftExecReal2Complex<float>(cufftHandle &plan,
+inline cufftResult cufftExecReal2Complex<float>(cufftHandle &plan,
                                          cufftReal_t<float> *d_in,
                                          cufftComplex_t<float> *d_out) {
   return cufftExecR2C(plan, d_in, d_out);
 }
 
 template <>
-cufftResult cufftExecReal2Complex<double>(cufftHandle &plan,
+inline cufftResult cufftExecReal2Complex<double>(cufftHandle &plan,
                                           cufftReal_t<double> *d_in,
                                           cufftComplex_t<double> *d_out) {
   return cufftExecD2Z(plan, d_in, d_out);
@@ -72,14 +72,14 @@ cufftResult cufftExecComplex2Real(cufftHandle &plan, cufftComplex_t<real> *d_in,
                                   cufftReal_t<real> *d_out);
 
 template <>
-cufftResult cufftExecComplex2Real<float>(cufftHandle &plan,
+inline cufftResult cufftExecComplex2Real<float>(cufftHandle &plan,
                                          cufftComplex_t<float> *d_in,
                                          cufftReal_t<float> *d_out) {
   return cufftExecC2R(plan, d_in, d_out);
 }
 
 template <>
-cufftResult cufftExecComplex2Real<double>(cufftHandle &plan,
+inline cufftResult cufftExecComplex2Real<double>(cufftHandle &plan,
                                           cufftComplex_t<double> *d_in,
                                           cufftReal_t<double> *d_out) {
   return cufftExecZ2D(plan, d_in, d_out);
@@ -99,14 +99,14 @@ cufftExecComplex2Complex(cufftHandle &plan, cufftComplex_t<real> *d_in,
                          cufftComplex_t<real> *d_out, int direction);
 
 template <>
-cufftResult
+inline cufftResult
 cufftExecComplex2Complex<float>(cufftHandle &plan, cufftComplex_t<float> *d_in,
                                 cufftComplex_t<float> *d_out, int direction) {
   return cufftExecC2C(plan, d_in, d_out, direction);
 }
 
 template <>
-cufftResult cufftExecComplex2Complex<double>(cufftHandle &plan,
+inline cufftResult cufftExecComplex2Complex<double>(cufftHandle &plan,
                                              cufftComplex_t<double> *d_in,
                                              cufftComplex_t<double> *d_out,
                                              int direction) {
diff --git a/src/utils/curandDebug.h b/src/utils/curandDebug.h
index 83c9e0a..653b215 100644
--- a/src/utils/curandDebug.h
+++ b/src/utils/curandDebug.h
@@ -7,7 +7,7 @@
 
 #define CurandSafeCall(err) __curandSafeCall(err, __FILE__, __LINE__)
 
-const char *curandGetErrorString(curandStatus_t err) {
+inline const char *curandGetErrorString(curandStatus_t err) {
   switch (err) {
   case CURAND_STATUS_VERSION_MISMATCH:
     return "CURAND_STATUS_VERSION_MISMATCH";
diff --git a/src/utils/cxx_utils.h b/src/utils/cxx_utils.h
index 07fb30a..bab80b0 100644
--- a/src/utils/cxx_utils.h
+++ b/src/utils/cxx_utils.h
@@ -29,7 +29,7 @@ inline std::string removePattern(std::string str, std::string pattern) {
 } // namespace stringUtils
 namespace printUtils {
 // Transform size in bytes to a pretty string in B, KB, MB...
-std::string prettySize(size_t size) {
+inline std::string prettySize(size_t size) {
   static const char *SIZES[] = {"B", "KB", "MB", "GB"};
   uint div = 0;
   size_t rem = 0;
diff --git a/src/utils/exception.h b/src/utils/exception.h
index 0425f77..fb78393 100644
--- a/src/utils/exception.h
+++ b/src/utils/exception.h
@@ -10,7 +10,7 @@
 namespace uammd {
 using exception = std::exception;
 
-void backtrace_nested_exception(const uammd::exception &e, int level = 0) {
+inline void backtrace_nested_exception(const uammd::exception &e, int level = 0) {
 
   Logging::log<Logging::EXCEPTION>(std::string(level, ' ') + "level " +
                                    std::to_string(level) +
